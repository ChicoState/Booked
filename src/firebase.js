import * as firebase from 'firebase';
import 'firebase/auth';

var config = {
  apiKey: "AIzaSyAbrPremBQusjYoKqKfTcVHwxuw-Q8ikHw",
  authDomain: "booked-8586e.firebaseapp.com",
  databaseURL: "https://booked-8586e.firebaseio.com",
  projectId: "booked-8586e",
  storageBucket: "booked-8586e.appspot.com",
  messagingSenderId: "24720035765",
  appId: "1:24720035765:web:3d57fd5c840b8100c353fc",
  measurementId: "G-20QKEY7V1D"
};

firebase.initializeApp(config);

// var provider = new firebase.auth.GoogleAuthProvider();
// //var user = firebase.auth().currentUser;
// var name, email;
//
//
// firebase.auth().useDeviceLanguage();
// firebase.auth().onAuthStateChanged(function(user){
//
// 	if(user != null){
// 		name = user.displayName;
// 		email = user.email;
// 		alert(name + ': ' + email);
// 	}
// 	else{
// 		alert('Working on Redirect for initial log in');
// 		firebase.auth.signinWithRedirect(provider);
// 		firebase.auth().getRedirectResult().then(function(result) {
// 			if (result.credential) {
// 				//This gives you a Google Access Token.
// 				//var token = result.credential.accessToken;
// 			}
// 			user = result.user;
// 		}).catch(function(error) {
// 				// Handle Errors here.
// 				var errorCode = error.code;
// 				var errorMessage = error.message;
// 				// The email of the user's account used.
// 				var errorEmail = error.email;
// 				// The firebase.auth.AuthCredential type that was used.
// 				var credential = error.credential;
//
// 				//Output line for error variables to silence unused var warnings
// 				alert(errorCode + errorMessage + errorEmail + credential);
// 		});
// 	}
// });
