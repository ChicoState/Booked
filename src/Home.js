import React from "react";
import app from "./firebase";
import firebase from "./firebase";
import 'firebase/database';

class Home extends React.Component {
  constructor(props) {
    super(props);
    this.state = { items: [], text: '', isSignedIn : false, hashItems: [], percentBooked: 0, noteSent: false };
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

loadDB() {
  const itemsRef = app.database().ref('text');
  let noteCheck = this.state.noteSent;
  itemsRef.on('value', (snapshot) => {
    let items = snapshot.val();
    let newState = [];
    let hashTags = [];
    let newHash = [];
    let uniqueHash = [];
    let datesBooked = [];
    if(firebase.auth().currentUser) {
      for (let item in items) {
        if(items[item].uid===firebase.auth().currentUser.uid) {
          var regexp = /\B\#\w\w+\b/g;
          let hashCheck = items[item].text.match(regexp);
          if (hashCheck !== null) {
            for(let tag in hashCheck) {
              hashTags.push(hashCheck[tag]);
            }
          }
          if(items[item].endDate==null) {
            items[item].endDate = items[item].startDate;
          }
          let startDCount = Math.floor((items[item].startDate - Date.now()) / (1000*60*60*24));
          let endDCount = Math.floor((items[item].endDate - Date.now()) / (1000*60*60*24));
          if (endDCount==0 && noteCheck==false) {
            noteCheck=true;
          }
          for(let i = 1; i < 31; i++) {
            if(i >= startDCount && i <= endDCount) {
              datesBooked.push(i);
            }
          }
          newState.push({
            id: items[item].id,
            text: items[item].text,
            creator: items[item].creator,
            key: item,
          });
        }
      }
      for (let item in items) {
        if(items[item].uid!==firebase.auth().currentUser.uid) {
          for(let tag in hashTags) {
            if(items[item].text.indexOf(hashTags[tag])!==-1) {
              newHash.push(items[item]);
            }
          }
        }
      }
    }
    uniqueHash = [...new Set(newHash)];
    datesBooked = [...new Set(datesBooked)];
    this.setState({
      items: newState,
      hashItems: uniqueHash,
      percentBooked: (Math.floor(datesBooked.length / 30 * 100)),
      noteSent: noteCheck
    });
  });
}

  componentDidMount = () => {

    firebase.auth().onAuthStateChanged(user => {
      this.setState({ isSignedIn: !!user})
      console.log("user", user)
    })
    if (!("Notification" in window)) {
      console.log("This browser does not support desktop notification");
    } else {
      Notification.requestPermission();
    }
    this.loadDB();
  }

  componentDidUpdate(prevProps,prevState) {
    if (this.state.isSignedIn !== prevState.isSignedIn) {
      this.loadDB();
      if(this.state.noteSent) {
        this.showNotification();
      }
    }
  }

  delete(id, key){
    this.setState({
      items: this.state.items.filter((x) => x.id != id )
    });
    app.database().ref('/text/' + key).remove();
  }


  // status(){
  //   firebase.auth().onAuthStateChanged(function(user){
  //     if(user)
  //     {
  //       console.log("currently signed in");
  //       firebase.auth().signOut();
  //     }
  //     else
  //     {
  //       console.log("not signed in");
  //       // var provider = new firebase.auth.GoogleAuthProvider();
  //       // firebase.auth().signInWithPopup(provider);
  //     }
  //   });
  // }

  render() {
    //let button = <button onClick={this.status}>TEST</button>
    const piestyle = {
        width: "75px",
        height: "75px",
//        borderRadius: "50%",
        background: "red",
        backgroundImage:"linear-gradient(to right, transparent " + this.state.percentBooked + "%, green 0)"
      }
    var notif_sound = "notif_sound";

    return (
      this.state.isSignedIn
      ?<div>
      <audio className="notifSound">
        <source src='/pen.mp3' type='audio/mpeg' />
      </audio>
        <h3>Kabooked Items</h3>
        <TodoList items={this.state.items} delete={this.delete.bind(this)} />
        <HashList items={this.state.hashItems} />
        <h3>Please Enter an Item to Kabook</h3>
        <form onSubmit={this.handleSubmit}>
          <label htmlFor="new-todo">
            Item Name:&nbsp;&nbsp;
          </label>
          <input
            id="new-todo"
            onChange={this.handleChange}
            value={this.state.text}
          />
          <button>
            Add #{this.state.items.length + 1}
          </button>
        </form>
        <br></br>
        (For Group Collaboration Place Shared #Hashtags Anywhere in Item Name)
        <h3>Next 30 Days</h3>
        <div style={piestyle}></div>
        <h3>{this.state.percentBooked} % Kabooked</h3>
      </div>
      :
        <div></div>


    );
  }

  showNotification() {
      var justdate = new Date();
      justdate.setHours(0,0,0,0);
      var options = {
        body: "You have Kabooked items ending soon!",
        tag:justdate
      };
     const audioNote = document.getElementsByClassName("notifSound")[0]
     let notification = new Notification("Kabooked Notification", options);
     notification.onshow = function() { audioNote.play(); };
//     notification.addEventListener("click", function() {
     //document.addEventListener("visibilitychange", function() {
//         audioNote.play();
//       });
//     }, false);
    }

  handleChange(e) {
    this.setState({ text: e.target.value });
  }

  handleSubmit(e) {
    e.preventDefault();
    if (this.state.text.length === 0) {
      return;
    }
    const itemsRef = app.database().ref('text');
    const newItem = {
      text: this.state.text,
      uid: firebase.auth().currentUser.uid,
      creator: firebase.auth().currentUser.displayName,
      groupColor: "blue",
      startDate: Date.now(),
      id: Date.now()
    };
    itemsRef.push(newItem);
  }
}

class TodoList extends React.Component {
  constructor() {
    super();
    this.onClickfn = this.onClickfn.bind(this);
  }

  onClickfn(id, key){
    this.props.delete(id, key);
  }
  render() {
    return (
      <ul>
        {this.props.items.map(item => (
          <li onClick={() => {if(window.confirm('Would you like to delete this item?')){this.onClickfn(item.id, item.key)};}} key={item.id}>{item.text}</li>
        ))}
      </ul>
    );
  }
}

class HashList extends React.Component {
  constructor() {
    super();
  }

  render() {
    return (
      (this.props.items!="")
      ?<ul>
        <h3>Other Tasks with Your Hashtags</h3>
        {this.props.items.map(item => (
          <li key={item.id}>{item.text} by {item.creator}</li>
        ))}
      </ul> : <ul></ul>
    );
  }
}

export default Home;
