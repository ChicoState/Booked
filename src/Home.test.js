import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import Home from './Home';
import FirebaseApp from './FirebaseApp'
import firebase from 'firebase';

const mockStore = configureMockStore([thunk]);
const store = mockStore();
const FirebaseAuthProps = {};

// jest.mock('firebase', () => {
//   return {
//     auth: () => {
//       return {
//         onAuthStateChanged: jest.fn(cb => cb({ name: 'Testy', uid: 'tGVt2mPFHXTEh9xl40SQEirzYaU2' }))
//       };
//     },
//   };
// });

// jest.mock('FirebaseApp', () => {
//   return mocksdk;
// });

// jest.mock('./FirebaseApp', () => {
//   const auth = jest.fn();
//   const mAuth = { signInWithRedirect: jest.fn() };
//   auth.GoogleAuthProvider = jest.fn();
//   auth.Auth = jest.fn(() => mAuth);
//   const data = { name: "text" };
//   const snapshot = { val: () => data };
//   auth.onAuthStateChanged = jest.fn(cb => cb({ name: 'Testy', uid: 'tGVt2mPFHXTEh9xl40SQEirzYaU2'}));
//   return {
//     auth,
//     initializeApp: jest.fn().mockReturnValue({
//       database: jest.fn().mockReturnValue({
//         ref: jest.fn().mockReturnThis(),
//         once: jest.fn(() => Promise.resolve(snapshot))
//       })
//     })
//   };
// });

// jest.mock('firebase', () => {
//   const auth = jest.fn();
//   const mAuth = { signInWithRedirect: jest.fn() };
//   auth.GoogleAuthProvider = jest.fn();
//   auth.Auth = jest.fn(() => mAuth);
//   const data = { name: "text" };
//   const snapshot = { val: () => data };
//   auth.onAuthStateChanged = jest.fn(cb => cb({ name: 'Testy', uid: 'tGVt2mPFHXTEh9xl40SQEirzYaU2'}));
//   return {
//     auth,
//     initializeApp: jest.fn().mockReturnValue({
//       database: jest.fn().mockReturnValue({
//         ref: jest.fn().mockReturnThis(),
//         once: jest.fn(() => Promise.resolve(snapshot))
//       })
//     })
//   };
// });





jest.mock('firebase', () => {
  const data = { name: "text" };
  const snapshot = { val: () => data };
  const ref = jest.fn().mockReturnThis();
  const on = jest.fn(() => Promise.resolve(snapshot));
  const database = jest.fn();
  database.Database = jest.fn().mockReturnValue({ref,on});

  const auth = jest.fn();
  const currentUser = jest.fn();

  const mAuth = { signInWithRedirect: jest.fn() };
  auth.currentUser = { displayName: 'Kris Selvidge', uid:'tGVt2mPFHXTEh9xl40SQEirzYaU2'};
  auth.GoogleAuthProvider = jest.fn();
  auth.Auth = jest.fn(() => mAuth);
  return { auth, database };
});

describe('<Home />', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });
const wrapper = mount(<Home />);
const wrapper2 = mount(<Home />);
const wrapper3 = mount(<Home />);
//wrapper.update();
if(wrapper.find('outty') > 0){
  wrapper.find('outty').simulate("click");
}







it('renders without crashing', () => {
  firebase.auth.mockImplementation(() => new firebase.auth.Auth());
  const div = document.createElement('div');
  ReactDOM.render(<Home />, div);
});

it('notifications', done => {
  firebase.auth.mockImplementation(() => new firebase.auth.Auth());
  const div = document.createElement('div');
  wrapper.setState({cuid:'7JmSHENe6vebq1H5JOP3n0VObsD3'});
  const itemsRef = FirebaseApp.database().ref('text');
  itemsRef.on('value', (snapshot) => {
    let items = snapshot.val();
     try {
       expect(items).anything;
       done();
     } catch (error) {
       done(error);
     }
  });
  wrapper.setState({cuid:'tGVt2mPFHXTEh9xl40SQEirzYaU2'});
  ReactDOM.render(wrapper, div);
  wrapper.update();
  wrapper.setState({noteSent:true});
  wrapper.setState({cuid:'7JmSHENe6vebq1H5JOP3n0VObsD3'});
  wrapper.update();
  wrapper.setState({isSignedIn:false});
  wrapper.setState({isSignedIn:true});
  ReactDOM.render(<Home />, div);
});

it('render with dataload', done => {
  firebase.auth.mockImplementation(() => new firebase.auth.Auth());
  const div = document.createElement('div');
  wrapper3.setState({cuid:'tGVt2mPFHXTEh9xl40SQEirzYaU2'});
  const itemsRef = FirebaseApp.database().ref('text');
  itemsRef.on('value', (snapshot) => {
    let items = snapshot.val();
     try {
       expect(items).anything;
       done();
     } catch (error) {
       done(error);
     }
  });
  wrapper3.setState({cuid:'tGVt2mPFHXTEh9xl40SQEirzYaU2'});
  ReactDOM.render(wrapper3, div);
  wrapper3.update();
  wrapper3.setState({cuid:'tGVt2mPFHXTEh9xl40SQEirzYaU2'});
  wrapper3.update();
  wrapper3.setState({isSignedIn:false});
  wrapper3.setState({isSignedIn:true});
  ReactDOM.render(<Home />, div);
});

it('render when signing in and out', done => {
  firebase.auth.mockImplementation(() => new firebase.auth.Auth());
  const div = document.createElement('div');
  wrapper2.setState({cuid:'tGVt2mPFHXTEh9xl40SQEirzYaU2'});
  const itemsRef = FirebaseApp.database().ref('text');
  itemsRef.on('value', (snapshot) => {
    let items = snapshot.val();
     try {
       expect(items).anything;
       done();
     } catch (error) {
       done(error);
     }
  });
  wrapper2.setState({cuid:'tGVt2mPFHXTEh9xl40SQEirzYaU2'});
  ReactDOM.render(wrapper2, div);
  wrapper2.update();
  wrapper2.setState({isSignedIn:true});
  wrapper2.setState({cuid:'tGVt2mPFHXTEh9xl40SQEirzYaU2'});
  wrapper2.update();
  wrapper2.setState({isSignedIn:false});
  wrapper2.setState({isSignedIn:true});
  ReactDOM.render(<Home />, div);
});


test('compare output', done => {
  firebase.auth.mockImplementation(() => new firebase.auth.Auth());
  firebase.database.mockImplementation(() => new firebase.database().Database());
  const itemsRef = FirebaseApp.database().ref('text');
  itemsRef.on('value', (snapshot) => {
    let items = snapshot.val();
     try {
       expect(items).anything;
       done();
     } catch (error) {
       done(error);
     }
  });
  const component = renderer.create(
    <Home />,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

});

// import React from 'react';
// import ReactDOM from 'react-dom';
// import renderer from 'react-test-renderer';
// import Home from './Home';
//


// it('renders without crashing', done => {
//   var testL;
//     const itemsRef = firebase.database().ref('text');
//   const itemsRef = FirebaseApp.database().ref('text').once('value').then((snapshot) => { testL = snapshot.val() || 'Nothing Here' });
//     console.log(testL);
//     console.log("Hi1");
//    itemsRef.on('value', (snapshot) => {
//      let items = snapshot.val();
//      try {
//        expect(items).anything;
//        done();
//      } catch (error) {
//        done(error);
//      }
//      console.log("Hi");
//      console.log(items);
//    });
//     console.log("HELLLLLLLLLO");
//
//   const div = document.createElement('div');
//   ReactDOM.render(<Home />, div);
// });
//
// test('compare output', () => {
//   const component = renderer.create(
//     <Home />,
//   );
//   let tree = component.toJSON();
//   expect(tree).toMatchSnapshot();
// });
