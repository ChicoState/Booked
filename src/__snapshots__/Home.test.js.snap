// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> compare output 1`] = `<div />`;

exports[`compare output 1`] = `<div />`;
